sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala

```diff
@@ -137,93 +113,61 @@ abstract class ColumnarOutputWriter(context: TaskAttemptContext,
-   * @return time in ns taken to write the batch
    */
-  private[this] def writeBatch(batch: ColumnarBatch): Long = {
-    if (includeRetry) {
-      writeBatchWithRetry(batch)
-    } else {
-      writeBatchNoRetry(batch)
-    }
-  }
-
-  /** Apply any necessary casts before writing batch out */
-  def transform(cb: ColumnarBatch): Option[ColumnarBatch] = None
-
-  private[this] def writeBatchWithRetry(batch: ColumnarBatch): Long = {
-    val sb = SpillableColumnarBatch(batch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-    RmmRapidsRetryIterator.withRetry(sb, RmmRapidsRetryIterator.splitSpillableInHalfByRows) { sb =>
-      val cr = new CheckpointRestore {
-        override def checkpoint(): Unit = ()
-        override def restore(): Unit = dropBufferedData()
-      }
-      val startTimestamp = System.nanoTime
-      withResource(sb.getColumnarBatch()) { cb =>
+  def writeSpillableAndClose(
+      spillableBatch: SpillableColumnarBatch,
+      statsTrackers: Seq[ColumnarWriteTaskStatsTracker]): Unit = {
+    val writeStartTime = System.nanoTime
+    val gpuTime = if (includeRetry) {
+      withRetry(spillableBatch, splitSpillableInHalfByRows) { sb =>
         //TODO: we should really apply the transformations to cast timestamps
         // to the expected types before spilling but we need a SpillableTable
         // rather than a SpillableColumnBatch to be able to do that
         // See https://github.com/NVIDIA/spark-rapids/issues/8262
-        RmmRapidsRetryIterator.withRestoreOnRetry(cr) {
-          withResource(new NvtxRange(s"GPU $rangeName write", NvtxColor.BLUE)) { _ =>
-            transform(cb) match {
-              case Some(transformed) =>
-                // because we created a new transformed batch, we need to make sure we close it
-                withResource(transformed) { _ =>
-                  scanAndWrite(transformed)
-                }
-              case _ =>
-                scanAndWrite(cb)
-            }
-          }
-        }
+        bufferBatchAndClose(sb.getColumnarBatch())
+      }.sum
+    } else {
+      withResource(spillableBatch) { _ =>
+        bufferBatchAndClose(spillableBatch.getColumnarBatch())
       }
-      GpuSemaphore.releaseIfNecessary(TaskContext.get)
-      val gpuTime = System.nanoTime - startTimestamp
-      writeBufferedData()
-      gpuTime
-    }.sum
+    }
+    // we successfully buffered to host memory, release the semaphore and write
+    // the buffered data to the FS
+    GpuSemaphore.releaseIfNecessary(TaskContext.get)
+    writeBufferedData()
+    updateStatistics(writeStartTime, gpuTime, statsTrackers)
   }
 
-  private[this] def writeBatchNoRetry(batch: ColumnarBatch): Long = {
-    var needToCloseBatch = true
-    try {
-      val startTimestamp = System.nanoTime
+  private[this] def bufferBatchAndClose(batch: ColumnarBatch): Long = {
+    val startTimestamp = System.nanoTime
+    withRestoreOnRetry(checkpointRestore) {
```

Comment:
This function is really confusing.

1. We have withRestoreOnRetry inside a method that is not always called from a withRetry method.
2. For the code path where withRetry is called it looks like this method has side effects and they are not handled properly.

Could we move the withRestoreOnRetry out to the calling code?

Reply:
> Could we move the withRestoreOnRetry out to the calling code?

Yes makes sense. I moved the `withRestoreOnRetry` over to right after the `withRetry` call in the `writeSpillableAndClose` function. 

The side effect I wasn't accounting for I believe is flipping the flag `anythingWritten` to true. When this flag has been set to `true` (which happens before the call to `tableWriter.write`) it means we wont' buffer an empty batch on `close`.

Do we want to make sure `anythingWritten` stays set to false in the case where `tableWriter.write` fails? If so, I can just swap two lines. For reference: https://github.com/NVIDIA/spark-rapids/blob/branch-23.08/sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala#L241

Reply:
Added comment around `anythingWritten` as discussed. It is OK if we flip this and then fail because we will retry. If the task fails all together the output will be removed anyway, so it does not matter much that this state changed.